### PR TITLE
Delay formatting of some messages unless they're going to be printed

### PR DESCRIPTION
**Pull request by drewejohnson**
This new function lives inside `include/openmc/error.h` and checks if a message will be written by checking the current
verbosity level prior to formatting the message. This works very similar to python logging modules and the spdlog project, where the call to the logging function accepts a formattable message string and the format arguments separately, e.g.

```cpp
format_write_message(1, "Writing data for statepoint {}", point)
```

The function accepts any number and type of argument after the format message and passes directly to `fmt::format` and in-turn to `openmc::write_message`. While this later function does similar checking with the level and MPI rank, the real goal is the output function in `src/error.cpp` which is not currently exposed through the error header file.

I also tried to reduce the number of string concatenations like `"The file " + file_path + " does not exist"` by using `fmt::format`. There are still some more places outside this PR, but these came up when I was modifying calls to `write_message` in favor of `format_write_message`.

### CODE REVIEW

File: include/openmc/error.h [hunk 1]
```diff
@@ -63,6 +66,13 @@ void write_message(const std::stringstream& message, int level)
   write_message(message.str(), level);
 }
 
+template <typename...Params>
+void format_write_message(const int level, const std::string& fmt_msg, const Params&...fmt_args) {
```
Comment (paulromano) [inline include/openmc/error.h:70]: `const` is superfluous on an integer passed by value
```suggestion
void format_write_message(int level, const std::string& fmt_msg, const Params&...fmt_args) {
```

Comment (drewejohnson) [reply include/openmc/error.h:70]: Thanks for this insight. I've added it to the updated version and will be pushing soon

**Comment by drewejohnson** [comment]
Thank you for the feedback @paulromano, the PR has been updated accordingly. I went with the different function name to specify a different intent, but overloading `write_message` feels a lot better and consistent :+1:

File: src/nuclide.cpp [hunk 2]
```diff
@@ -944,7 +944,7 @@ extern "C" int openmc_load_nuclide(const char* name, const double* temps, int n)
     // Get filename for library containing nuclide
     int idx = it->second;
     const auto& filename = data::libraries[idx].path_;
-    write_message("Reading " + std::string{name} + " from " + filename, 6);
+    write_message(6, "Reading {} from {}", std::string{name}, filename);
```
Comment (paulromano) [inline src/nuclide.cpp:947]:
```suggestion
    write_message(6, "Reading {} from {}", name, filename);
```

File: src/particle.cpp [hunk 3]
```diff
@@ -410,7 +410,7 @@ Particle::cross_surface()
   // TODO: off-by-one
   const auto& surf {model::surfaces[i_surface - 1].get()};
   if (settings::verbosity >= 10 || trace_) {
-    write_message("    Crossing surface " + std::to_string(surf->id_));
+    write_message(fmt::format("    Crossing surface {}", surf->id_));
```
Comment (paulromano) [inline src/particle.cpp:413]:
```suggestion
    write_message("    Crossing surface {}", surf->id_);
```

File: src/particle.cpp [hunk 4]
```diff
@@ -437,7 +437,7 @@ Particle::cross_surface()
 
     // Display message
     if (settings::verbosity >= 10 || trace_) {
-      write_message("    Leaked out of surface " + std::to_string(surf->id_));
+      write_message(fmt::format("    Leaked out of surface {}", surf->id_));
```
Comment (paulromano) [inline src/particle.cpp:440]:
```suggestion
      write_message("    Leaked out of surface {}", surf->id_);
```

File: src/particle.cpp [hunk 5]
```diff
@@ -502,7 +502,7 @@ Particle::cross_surface()
 
     // Diagnostic message
     if (settings::verbosity >= 10 || trace_) {
-      write_message("    Reflected from surface " + std::to_string(surf->id_));
+      write_message(fmt::format("    Reflected from surface {}", surf->id_));
```
Comment (paulromano) [inline src/particle.cpp:505]:
```suggestion
      write_message("    Reflected from surface {}", surf->id_);
```

File: src/particle.cpp [hunk 6]
```diff
@@ -556,8 +556,8 @@ Particle::cross_surface()
 
     // Diagnostic message
     if (settings::verbosity >= 10 || trace_) {
-      write_message("    Hit periodic boundary on surface " +
-        std::to_string(surf->id_));
+      write_message(fmt::format("    Hit periodic boundary on surface {}",
+        surf->id_));
```
Comment (paulromano) [inline src/particle.cpp:560]:
```suggestion
      write_message("    Hit periodic boundary on surface {}", surf->id_);
```

File: src/simulation.cpp [hunk 7]
```diff
@@ -500,7 +501,7 @@ void initialize_history(Particle& p, int64_t index_source)
 
   // Display message if high verbosity or trace is on
   if (settings::verbosity >= 9 || p.trace_) {
-    write_message("Simulating Particle " + std::to_string(p.id_));
+    write_message(fmt::format("Simulating Particle {}", p.id_));
```
Comment (paulromano) [inline src/simulation.cpp:504]:
```suggestion
    write_message("Simulating Particle {}", p.id_);
```

**Pull request merged**